domain/stub/service.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package stub
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	jujuerrors "github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/internal/errors"
+)
+
+// StubService is a special service that collects temporary methods required for
+// wiring together domains which not completely implemented or wired up.
+//
+// Given the temporary nature of this service, we have not implemented the full
+// service/state layer indirection. Instead, the service directly uses a transaction
+// runner.
+//
+// All methods here should be thrown away as soon as we're done with then.
+type StubService struct {
+	*domain.StateBase
+}
+
+// NewStubService returns a new StubService.
+func NewStubService(factory database.TxnRunnerFactory) *StubService {
+	return &StubService{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func (s *StubService) AssignUnitsToMachines(ctx context.Context, groupedUnitsByMachine map[string][]string) error {
+	db, err := s.DB()
+	if err != nil {
+		return jujuerrors.Trace(err)
+	}
+
+	getNetNodeQuery, err := s.Prepare(`
+SELECT &netNodeUUID.*
+FROM machine
+WHERE name = $machine.name
+`, netNodeUUID{}, machine{})
+	if err != nil {
+		return errors.Errorf("failed to prepare machine query: %v", err)
+	}
+
+	setUnitNetNodeQuery, err := s.Prepare(`
+UPDATE unit
+SET net_node_uuid = $netNodeUUID.net_node_uuid
+WHERE name = $unit.name
+`, netNodeUUID{}, unit{})
+	if err != nil {
+		return errors.Errorf("failed to prepare unit query: %v", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		i := 0
+		for machine, units := range encodeGroupedUnitsByMachine(groupedUnitsByMachine) {
+			var netNodeUUID netNodeUUID
+			err = tx.Query(ctx, getNetNodeQuery, machine).Get(&netNodeUUID)
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Errorf("machine not found: %v", machine.MachineName)
+			}
+			if err != nil {
+				return errors.Errorf("failed to get machine net node: %v", err)
+			}
+
+			for _, unit := range units {
+				var outcome sqlair.Outcome
+				err = tx.Query(ctx, setUnitNetNodeQuery, unit, netNodeUUID).Get(&outcome)
+				if err != nil {
+					return errors.Errorf("failed to set unit: %v", err)
+				}
+				rowsEffected, err := outcome.Result().RowsAffected()
+				if err != nil {
+					return errors.Errorf("failed to get unit rows effected: %v", err)
+				}
+				if rowsEffected == 0 {
+					return errors.Errorf("unit not found: %v", unit.UnitName)
+				}
```

Comment:
This should return a unit not found error, not just a string error. 

I'd also like to check the unit exists, before we attempt the update, then you can slim down what you actually need to update, or return an error early.

Also, this method is inefficient, but because it's only on the stub service I can let that fly. Can we add a comment to that fact.

Reply:
Yeah if I was writing production code I would do the set in a batch. However I'm not worried because it's only a stub.

Reply:
I improved it anyway